### PR TITLE
Add more counters for scrapes: attempts, success, errors and timeouts

### PR DESCRIPTION
**Pull request by vii**
It is currently hard to monitor if exporters timeout because of long collections.

This adds more per scrape counters to allow this to be monitored.

It might be useful to also track total scrape time spent.

### CODE REVIEW

**Comment by brian-brazil** [comment]
`up` should cover this already, and at a more granular level.

**Comment by vii** [comment]
appreciate the quick response @brian-brazil 

The issue we have is that some of our exporters gradually report more and more metrics, and therefore slow down. The timeout condition is quite different from the host being 'down'. 

What do you think is the best way to alert on this before the host gets so slow it is actually down?

I could remove the 'error' and 'success' counters if you don't think they're useful and add a 'last successful scrape duration' gauge, so that it would be possible to alert on the exporters slowing down. Alternatively, a 'total time spent in scrapes' for a target would also be a useful metric.

Let me know what you suggest -

**Comment by brian-brazil** [comment]
> What do you think is the best way to alert on this before the host gets so slow it is actually down?

There's also `scrape_duration_seconds`, though if you're seeing this with direct instrumentation using a client library then something is very wrong.

**Comment by vii** [comment]
This actually can happen for the JMX exporter, on a variety of services, for example, if GC hits. For various batch bulk jobs long GC pauses may not be entirely undesirable. In addition, any performance bug in a JVM exposed property would lead to this.

It would be useful for us to be able to distinguish that a timeout occurred, rather than some other error. Happy to simplify the pull request -

**Comment by brian-brazil** [comment]
This PR will not help you determine that as the metrics are too broad to be useful for this. The existing metrics are as granular as we can sanely get on that.

**Comment by matthiasr** [comment]
I disagree that the existing metrics are sufficient. I always need to go to the target page, find the target, read the message, and make an educated guess as to what went wrong. I would like to be able to tell _why_ it failed based on metrics too.

Slow scrapes happen, and there are legitimate use cases where the scrape duration depends on circumstances beyond the control of the scraped software.

The `scrape_duration_seconds` is _not_ sufficient, if I were to propose a "last measurement" gauge for latency anywhere else you'd throw something at me, most likely a histogram.

**Comment by brian-brazil** [comment]
> I would like to be able to tell why it failed based on metrics too.

That's a use case for logging, not metrics.

> The scrape_duration_seconds is not sufficient, if I were to propose a "last measurement" gauge for latency anywhere else you'd throw something at me, most likely a histogram.

This is okay in this case as it relates to the scrape itself, it's the same as pushing batch job duration to the pushgateway or the snmp/blackbox exporters reporting their scrape latencies.

File: retrieval/scrape.go [hunk 1]
```diff
@@ -53,6 +59,24 @@ var (
...
+	)
+	targetSuccessfulScrapes = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "prometheus_target_successful_scrapes_total",
+			Help: "Total number of scrapes that were successful.",
+		},
+	)
+	targetTimeoutScrapes = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "prometheus_target_timeout_scrapes_total",
```
Comment (vii) [inline retrieval/scrape.go]: this is the key new metric which allows the people to see when a scrape times out, as opposed to another kind of error

File: retrieval/scrape.go [hunk 2]
```diff
@@ -421,8 +450,14 @@ func (sl *scrapeLoop) run(interval, timeout time.Duration, errc chan<- error) {
 
 			samples, err := sl.scraper.scrape(scrapeCtx, start)
 			if err == nil {
+				targetSuccessfulScrapes.Inc()
 				sl.append(samples)
 			} else if errc != nil {
+				if err == context.DeadlineExceeded {
+					targetTimeoutScrapes.Inc()
```
Comment (vii) [inline retrieval/scrape.go]: and here is where it is incremented

**Comment by vii** [comment]
the extra metric "prometheus_target_timeout_scrapes_total" shows this particular case, which is quite interesting to alert on - it indicates that sometimes the metrics exporter is too slow, and that the scrape timeout should be increased

This is hard to see from scrape_duration, as that doesn't indicate whether it fails, and up might legitimately be 0 when the service is down. Having metrics stuck (taking longer than the timeout) but the service up, is a real world case we have observed in several daemons

**Comment by brian-brazil** [comment]
You're still not proposing anything that's not already possible with existing generated metrics.

**Comment by discordianfish** [comment]
Agree with @matthiasr and @vii, have a metric which splits out the reason for a failed scrape would be very useful. Scrape duration split out by resolve/connect/read would be useful too.

**Comment by brian-brazil** [comment]
https://github.com/prometheus/prometheus/issues/2820 covers providing this information.

**Pull request closed without merging**